kyverno-test-util.py

```diff
@@ -0,0 +1,132 @@
+    """
+    resources = list()
+
+    for root, dirs, files in os.walk(f'{folder}'):
+        for file in files:
+            if (file.endswith(".yaml") or file.endswith(".yml")) and ("kyverno" not in root) and len([ignored for ignored in ignoredFileNames if (file.find(ignored) != -1)]) == 0:
+                resources.append(os.path.join(root, file))
+    return resources
+
+def writeResources(folder):
+    resources = {'resources': getResourceLocations(folder)}
+    appendToFile(resources)
+
+def writeResults(folder, policies_location):
+    results = {'results': getPreparedResults(folder, policies_location)}
+    appendToFile(results)
+
+def getExpectedResult(rule, resource):
+    """
+    This method is meant to determine what should be the result of given test.
+    We don't have tests that are meant to fail, only pass, or if resource is to be ignored - skip.
+    """
+    shouldExclude = 0
+    if 'exclude' in rule.keys():
+        if 'namespace' in str(rule['exclude']) and resource['metadata']['namespace'] in rule['exclude']['any'][0]['resources']['namespaces']:
+            shouldExclude = 1
+
+        if 'kinds' in str(rule['exclude']) and resource['metadata']['name'] in rule['exclude']['any'][0]['resources']['names']:
+            shouldExclude = 1
+
+    match shouldExclude:
+        case 0:
+            return 'pass'
+        case 1:
+            return 'skip'
+
+def getResult(policy, rule, resource):
+    result = dict()
+    if resource['kind'] in allowedResources:
+        result = {
+            'policy': policy['metadata']['name'],
+            'rule': rule['name'],
+            'resources': [resource['metadata']['name']],
+            'result': getExpectedResult(rule, resource),
+            'kind': resource['kind']
+        }
+
+        if 'namespace' in resource['metadata'].keys():
+            result['namespace'] = resource['metadata']['namespace']
+
+    return result
+
+def getPreparedResults(folder, policies_location):
+    """
+    Because there are multiple exclusions in policies and resources can have multiple objects in one file,
+    this method iterates over everyting and strives to build `result` element for every available policy rule - resource combination
+    """
+    resourceLocations = getResourceLocations(folder)
+    results = list()
+    with open(policies_location) as policyFile:
+        policies = yaml.safe_load_all(policyFile)
+        for policy in policies:
+            for rule in policy['spec']['rules']:
+                for resource in resourceLocations:
+                    with open(resource) as file:
+                        resources = yaml.safe_load_all(file)
+                        for resource in resources:
+                            if type(resource) is dict:
+                                result = getResult(policy, rule, resource)
+                            else:
+                                for obj in resource:
```

Comment:
I sometimes get stuck here because `resource` is `None`.
I don't know exactly why, but for the time I fixed it by changing this `else:` to `elif resource:`

> Traceback (most recent call last):
>   File "kyverno-test-util.py", line 133, in <module>
>     main()
>   File "kyverno-test-util.py", line 130, in main
>     writeResults(resources_folder, policies_location)
>   File "kyverno-test-util.py", line 42, in writeResults
>     results = {'results': getPreparedResults(folder, policies_location)}
>   File "kyverno-test-util.py", line 99, in getPreparedResults
>     for obj in resource:
> TypeError: 'NoneType' object is not iterable
>